client/peerset/src/lib.rs

```diff
@@ -199,6 +208,8 @@ pub struct Peerset {
 	created: Instant,
 	/// Last time when we updated the reputations of connected nodes.
 	latest_time_update: Instant,
+	/// List of nodes that are allowed to connect.
+	allowlist: Option<Vec<PeerId>>,
```

Comment:
As discussed, I'm really not a fan of adding this. The way the peerset works is already very obscure, and I'd prefer avoiding adding yet another mechanism to it. We already have reserved nodes and priority groups.
It's unclear to me exactly what to do here (cc #6087), but for your initial prototype you can just use `set_reserved_nodes` instead of `set_allowlist`.